File: src/lib/last-seen-relative-ru.ts
import { differenceInCalendarDays, differenceInMonths, startOfDay } from 'date-fns';

const PREFIX = 'Был(а) ';

/** 11–14 → родительный множественный (минут, часов, …). */
function isTeen(n: number): boolean {
  const m = n % 100;
  return m >= 11 && m <= 14;
}

function ruMinutesPhrase(n: number): string {
  if (n <= 0) return 'менее минуты назад';
  if (n === 1) return 'минуту назад';
  if (isTeen(n)) return `${n} минут назад`;
  const t = n % 10;
  if (t === 1) return `${n} минуту назад`;
  if (t >= 2 && t <= 4) return `${n} минуты назад`;
  return `${n} минут назад`;
}

function ruHoursPhrase(n: number): string {
  if (n <= 0) return 'менее часа назад';
  if (n === 1) return 'час назад';
  if (isTeen(n)) return `${n} часов назад`;
  const t = n % 10;
  if (t === 1) return `${n} час назад`;
  if (t >= 2 && t <= 4) return `${n} часа назад`;
  return `${n} часов назад`;
}

function ruDaysPhrase(n: number): string {
  if (isTeen(n)) return `${n} дней назад`;
  const t = n % 10;
  if (t === 1) return `${n} день назад`;
  if (t >= 2 && t <= 4) return `${n} дня назад`;
  return `${n} дней назад`;
}

/** Именительный для состава «X лет Y месяцев». */
function ruMonthsNominal(n: number): string {
  if (isTeen(n)) return `${n} месяцев`;
  const t = n % 10;
  if (t === 1) return `${n} месяц`;
  if (t >= 2 && t <= 4) return `${n} месяца`;
  return `${n} месяцев`;
}

function ruYearsNominal(n: number): string {
  if (isTeen(n)) return `${n} лет`;
  const t = n % 10;
  if (t === 1) return `${n} год`;
  if (t >= 2 && t <= 4) return `${n} года`;
  return `${n} лет`;
}

/**
 * Строка статуса «последняя активность» для профиля (относительное время по-русски).
 * Примеры: «Был(а) менее минуты назад», «Был(а) минуту назад», «Был(а) вчера», «Был(а) 1 год 7 месяцев назад».
 */
export function formatLastSeenStatusRu(lastSeen: Date, now: Date = new Date()): string {
  let diffMs = now.getTime() - lastSeen.getTime();
  if (diffMs < 0) diffMs = 0;

  const sec = Math.floor(diffMs / 1000);
  const min = Math.floor(diffMs / 60000);

  if (sec < 60) {
    return PREFIX + 'менее минуты назад';
  }
  if (min < 60) {
    return PREFIX + ruMinutesPhrase(min);
  }

  const calDays = differenceInCalendarDays(startOfDay(now), startOfDay(lastSeen));

  if (calDays === 0) {
    const hrs = Math.max(1, Math.floor(diffMs / 3600000));
    return PREFIX + ruHoursPhrase(hrs);
  }

  if (calDays === 1) {
    return PREFIX + 'вчера';
  }

  if (calDays >= 2 && calDays <= 30) {
    return PREFIX + ruDaysPhrase(calDays);
  }

  const totalMonths = differenceInMonths(now, lastSeen);

  if (totalMonths < 1) {
    return PREFIX + ruDaysPhrase(calDays);
  }

  if (totalMonths < 12) {
    if (totalMonths === 1) return PREFIX + '1 месяц назад';
    return PREFIX + `${ruMonthsNominal(totalMonths)} назад`;
  }

  const years = Math.floor(totalMonths / 12);
  const monthsRem = totalMonths % 12;

  if (monthsRem === 0) {
    return PREFIX + `${ruYearsNominal(years)} назад`;
  }

  return PREFIX + `${ruYearsNominal(years)} ${ruMonthsNominal(monthsRem)} назад`;
}

File: mobile/app/lib/features/chat/data/last_seen_relative_ru.dart
/// Относительная строка «Был(а) …» для lastSeen (как на вебе).
String formatLastSeenStatusRu(DateTime lastSeen, [DateTime? now]) {
  final n = now ?? DateTime.now();
  var diffMs = n.millisecondsSinceEpoch - lastSeen.millisecondsSinceEpoch;
  if (diffMs < 0) diffMs = 0;

  final sec = diffMs ~/ 1000;
  final min = diffMs ~/ 60000;

  if (sec < 60) return '$_prefixменее минуты назад';
  if (min < 60) return '$_prefix${_ruMinutesPhrase(min)}';

  final calDays = _differenceInCalendarDays(n, lastSeen);

  if (calDays == 0) {
    final hrs = (diffMs ~/ 3600000).clamp(1, 999999);
    return '$_prefix${_ruHoursPhrase(hrs)}';
  }

  if (calDays == 1) return '$_prefixвчера';

  if (calDays >= 2 && calDays <= 30) {
    return '$_prefix${_ruDaysPhrase(calDays)}';
  }

  final totalMonths = _differenceInMonths(n, lastSeen);

  if (totalMonths < 1) {
    return '$_prefix${_ruDaysPhrase(calDays)}';
  }

  if (totalMonths < 12) {
    if (totalMonths == 1) return '${_prefix}1 месяц назад';
    return '$_prefix${_ruMonthsNominal(totalMonths)} назад';
  }

  final years = totalMonths ~/ 12;
  final monthsRem = totalMonths % 12;

  if (monthsRem == 0) {
    return '$_prefix${_ruYearsNominal(years)} назад';
  }

  return '$_prefix${_ruYearsNominal(years)} ${_ruMonthsNominal(monthsRem)} назад';
}

const _prefix = 'Был(а) ';

bool _isTeen(int n) {
  final m = n % 100;
  return m >= 11 && m <= 14;
}

int _differenceInCalendarDays(DateTime a, DateTime b) {
  final startA = DateTime(a.year, a.month, a.day);
  final startB = DateTime(b.year, b.month, b.day);
  return startA.difference(startB).inDays;
}

int _differenceInMonths(DateTime end, DateTime start) {
  var months = (end.year - start.year) * 12 + (end.month - start.month);
  if (end.day < start.day) months--;
  return months;
}

String _ruMinutesPhrase(int n) {
  if (n <= 0) return 'менее минуты назад';
  if (n == 1) return 'минуту назад';
  if (_isTeen(n)) return '$n минут назад';
  final t = n % 10;
  if (t == 1) return '$n минуту назад';
  if (t >= 2 && t <= 4) return '$n минуты назад';
  return '$n минут назад';
}

String _ruHoursPhrase(int n) {
  if (n <= 0) return 'менее часа назад';
  if (n == 1) return 'час назад';
  if (_isTeen(n)) return '$n часов назад';
  final t = n % 10;
  if (t == 1) return '$n час назад';
  if (t >= 2 && t <= 4) return '$n часа назад';
  return '$n часов назад';
}

String _ruDaysPhrase(int n) {
  if (_isTeen(n)) return '$n дней назад';
  final t = n % 10;
  if (t == 1) return '$n день назад';
  if (t >= 2 && t <= 4) return '$n дня назад';
  return '$n дней назад';
}

String _ruMonthsNominal(int n) {
  if (_isTeen(n)) return '$n месяцев';
  final t = n % 10;
  if (t == 1) return '$n месяц';
  if (t >= 2 && t <= 4) return '$n месяца';
  return '$n месяцев';
}

String _ruYearsNominal(int n) {
  if (_isTeen(n)) return '$n лет';
  final t = n % 10;
  if (t == 1) return '$n год';
  if (t >= 2 && t <= 4) return '$n года';
  return '$n лет';
}
